### DISCUSSION TITLE
How to use Controllers from other projects?

### DISCUSSION THREAD

**Discussion by hieucd04** [OrchardCMS/OrchardCore#14410] [Q&A] [answered]
I have a normal .NET project with a bunch of controllers in it (let's call it `Api Project`). I don't want to modify the `Api Project`. So, I created a new project, turned it into an Orchard Core module and then had it reference the `Api Project` with `PrivateAssets='none'`. But it doesn't work. None of the controllers get registered or called.

If I convert the `Api Project` into an Orchard Core module then all controllers get registered and work fine. However, like I said, I really don't want to modify the `Api Project`.

Is there anyway to register or call controllers which live outside of an Orchard Core module?

**Comment by Piedone** [comment]
You can try using [`AddControllers()`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.mvcservicecollectionextensions.addcontrollers?view=aspnetcore-7.0) in your module's `Startup` class, though I'm not sure if and how it'll work, but it's worth a try.

Also, is it not feasible to create controllers in the module and use services from `Api Project` instead?

**Comment by hieucd04** [comment]
> You can try using [AddControllers()](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.mvcservicecollectionextensions.addcontrollers?view=aspnetcore-7.0) in your module's Startup class

I tried it. Unfortunately, it didn't work!

I also tried using Application Part Manager as suggested here: https://stackoverflow.com/questions/47617994/how-to-use-a-controller-in-another-assembly-in-asp-net-core-mvc-2-0. But it also didn't work.
```csharp
public void ConfigureServices(IServiceCollection services)
{
    var assembly = typeof(**AnyTypeFromRequiredAssembly**).Assembly;

    services.AddControllers()
        .PartManager.ApplicationParts.Add(new AssemblyPart(assembly));
}
```

Another attempt as suggested here: https://github.com/OrchardCMS/OrchardCore/issues/696. But still no result.
```csharp
public override void Configure(IApplicationBuilder app, IRouteBuilder routes, IServiceProvider serviceProvider)
    {
        var partManager = serviceProvider.GetRequiredService<ApplicationPartManager>();

        // by providing a type defined in the assembly
        //partManager.ApplicationParts.Add(new AssemblyPart(
        //    typeof(MyClH1TagHelper).GetTypeInfo().Assembly));

        // or by providing an assembly name
        // note: it's fast because the assembly is already loaded
        partManager.ApplicationParts.Add(new AssemblyPart(
            Assembly.Load(new AssemblyName("ClTagHelpers"))));
```
---

> Also, is it not feasible to create controllers in the module and use services from Api Project instead?

Well, there are all kind of permissions & "who is allowed to do what" in a company. Our team don't have permissions to submit PR to the `API Project` only Pull from it. Also, there are already Controllers inside the `API Project`. So, we really don't want to repeat ourself by creating / removing controllers every time the `API Project` changes.

**Reply by Piedone** [reply]
I don't have further suggestions, sorry.

**Reply by jtkech** [reply, marked as answer]
Yes, only controller types defined in a module (not from an external library) can be used, can be tied to a module feature by using the `[Feature()]` attribute, otherwise tied to the main feature, the module itself if no feature is expllicitly defined.

https://github.com/OrchardCMS/OrchardCore/blob/40fa1346ef2f66d6d0398a95f9cc72814f6c6342/src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs#L62

What you could do is to define Wrappers, a given Wrapper would inherit from a Controller type defined in your library.

**Reply by hieucd04** [reply]
Based on your advice, we finally settled down to the solution below. Not the nicest one but it works in our case and we don't have to add/remove `Wrappers` every time the API project adds/removes Controllers.

```csproj
<ItemGroup>
  <!-- Link the entire API project into this Orchard Core module. So that all of it Controllers will be registered. -->
  <Compile
    Include="path\to\API\project\**"
    Exclude="$(DefaultItemExcludes);**\bin\**;**\obj\**;**\Properties\**;"
    LinkBase="_Upstream"
  />
</ItemGroup>

<ItemGroup>
  <!-- Aliases is added to avoid CS0436 warning. Since we don't use it, we just put a random GUID in. -->
  <ProjectReference Include="path\to\API\project" Aliases="c3d57711ab0f416c912f86afebb0d79f" />
</ItemGroup>

<ItemGroup>
  <!-- This package is required to turn this project into an Orchard Core module. -->
  <PackageReference Include="OrchardCore.Module.Targets" />
</ItemGroup>
```

**Reply by jtkech** [reply]
Cool!
